packages/augur-core/src/contracts/para/AMMExchange.sol

```diff
@@ -8,236 +8,251 @@ import 'ROOT/ICash.sol';
-}
\ No newline at end of file
+    using SafeMathUint256 for uint256;
+
+    ICash public cash;
+    IParaShareToken public shareToken;
+    IMarket public augurMarket;
+    uint256 public numTicks;
+    uint256 public INVALID;
+    uint256 public NO;
+    uint256 public YES;
+    uint256 public fee; // [0-1000] how many thousandths of swaps should be kept as fees
+    uint256 private liquidityConstant; // defaults to zero
+
+    constructor(IMarket _market, IParaShareToken _shareToken, uint256 _fee) public {
+        cash = _shareToken.cash();
+        shareToken = _shareToken;
+        augurMarket = _market;
+        numTicks = _market.getNumTicks();
+        cash.approve(address(_shareToken), 2**256-1);
+        cash.approve(address(_shareToken.augur()), 2**256-1);
+        INVALID = _shareToken.getTokenId(_market, 0);
+        NO = _shareToken.getTokenId(_market, 1);
+        YES = _shareToken.getTokenId(_market, 2);
+        fee = _fee;
+    }
+
+    // Adds shares to the liquidity pool by minting complete sets.
+    function addLiquidity(uint256 _sharesToBuy) external {
+        cash.transferFrom(msg.sender, address(this), _sharesToBuy.mul(numTicks));
+        shareToken.publicBuyCompleteSets(augurMarket, _sharesToBuy);
+
+        uint256 _newLiquidityConstant = calculateLiquidityConstant();
+
+        if (liquidityConstant == 0) {
+            // The formula for issuing LP tokens breaks down when the supply (and therefore liquidityConstant) are zero.
+            _mint(msg.sender, _newLiquidityConstant);
+        } else {
+            uint256 _totalSupply = totalSupply;
+            // User gains LP tokens relative to how many LP tokens were already issued, and the change in the ratio of shares in the pool.
+            // User gains more LP tokens if they are an earlier
+            _mint(msg.sender, _totalSupply.mul(_newLiquidityConstant).div(liquidityConstant).sub(_totalSupply));
+        }
+
+        liquidityConstant = _newLiquidityConstant;
+    }
+
+    // Removes shares from the liquidity pool; does not redeem complete sets for cash.
+    function removeLiquidity(uint256 _poolTokensToSell) external {
+        uint256 _poolSupply = totalSupply;
+        (uint256 _poolInvalid, uint256 _poolNo, uint256 _poolYes) = shareBalances(address(this));
+        uint256 _poolCash = cash.balanceOf(address(this));
+        uint256 _invalidShare = _poolInvalid.mul(_poolTokensToSell).div(_poolSupply);
+        uint256 _noShare = _poolNo.mul(_poolTokensToSell).div(_poolSupply);
+        uint256 _yesShare = _poolYes.mul(_poolTokensToSell).div(_poolSupply);
+        uint256 _cashShare = _poolCash.mul(_poolTokensToSell).div(_poolSupply);
+        _burn(msg.sender, _poolTokensToSell);
+        shareTransfer(address(this), msg.sender, _invalidShare, _noShare, _yesShare);
+        cash.transfer(msg.sender, _cashShare);
+        liquidityConstant = calculateLiquidityConstant();
```

Comment:
Why is this a storage variable?  Seems like it can be a local calculated from pool balance.  Perhaps it is an attempt at gas savings?

Reply:
It is. I figured it would be cheaper if adding liquidity were a bit more common than removing liquidity. But today I learned storage will become significantly more costly. I'll switch back to calculating each time.

Reply:
Storage *reads* are going to be come more expensive, so this optimization may still be valuable if it minimizes the number of reads.

Reply:
It adds a storage read actually so it's just worse. Switching to calculating each time.